### PR TITLE
QUIC: Refresh ack frame when previous ack_frame lost

### PR DESCRIPTION
**Pull request by scw00**
>13.2. Retransmission of Information
>The most recent set of acknowledgments are sent in ACK frames. An ACK frame SHOULD contain all unacknowledged acknowledgments, as described in Section 13.1.1.

### CODE REVIEW

File: iocore/net/quic/QUICAckFrameCreator.cc [hunk 1]
```diff
@@ -104,6 +104,15 @@ QUICAckFrameManager::_on_frame_acked(QUICFrameInformationUPtr &info)
   this->_ack_creator[index]->forget(ack_info->largest_acknowledged);
 }
 
+void
+QUICAckFrameManager::_on_frame_lost(QUICFrameInformationUPtr &info)
+{
+  ink_assert(info->type == QUICFrameType::ACK);
+  int index = QUICTypeUtil::pn_space_index(info->level);
+  // when ack frame lost. Force to refresh the frame.
+  this->_ack_creator[index]->refresh_frame();
```
Comment (maskit) [inline iocore/net/quic/QUICAckFrameCreator.cc]: We're not refreshing *frame* anymore. Maybe `refresh_state`?

File: iocore/net/quic/QUICAckFrameCreator.cc [hunk 2]
```diff
@@ -116,27 +125,20 @@ QUICAckFrameManager::ack_delay_exponent() const
...
 //
 // QUICAckFrameManager::QUICAckFrameCreator
 //
 void
 QUICAckFrameManager::QUICAckFrameCreator::refresh_frame()
 {
-  if (this->_available) {
-    // make sure we have the new ack_frame to override the old one.
-    this->_ack_frame = this->create_ack_frame();
+  QUICConfig::scoped_config params;
```
Comment (maskit) [inline iocore/net/quic/QUICAckFrameCreator.cc]: Do we need this?

File: iocore/net/quic/QUICAckFrameCreator.cc (hunk 1 above)
Comment (scw00) [reply iocore/net/quic/QUICAckFrameCreator.cc]: will do .

File: iocore/net/quic/QUICAckFrameCreator.cc (hunk 2 above)
Comment (scw00) [reply iocore/net/quic/QUICAckFrameCreator.cc]: will do .

File: iocore/net/quic/QUICAckFrameCreator.cc [hunk 3]
```diff
@@ -358,9 +343,17 @@ QUICAckFrameManager::QUICAckFrameCreator::available() const
...
 
 bool
-QUICAckFrameManager::QUICAckFrameCreator::is_ack_frame_ready() const
+QUICAckFrameManager::QUICAckFrameCreator::is_ack_frame_ready()
 {
-  return this->_ack_frame != nullptr;
+  QUICConfig::scoped_config params;
+
+  if (this->_available && this->_has_new_data && !this->_packet_numbers.empty() &&
+      this->_latest_packet_received_time + params->max_ack_delay_in() <= Thread::get_hrtime()) {
```
Comment (maskit) [inline iocore/net/quic/QUICAckFrameCreator.cc]: `max_ack_delay_in` will be sent as Transport Parameter and it won't be changed during a session. We should read it at initialization and keep it in a member variable.

**Comment by maskit** [comment]
Overall, looks fine. I may want to merge this next week because we are on interop this week.

**Comment by scw00** [comment]
Ok,

**Comment by scw00** [comment]
updated.

File: iocore/net/quic/QUICAckFrameCreator.cc (hunk 3 above)
Comment (scw00) [reply iocore/net/quic/QUICAckFrameCreator.cc]: I changed max_ack_delay_in to max_ack_delay_out . 

>The maximum ACK delay is an integer value indicating the maximum amount of time in milliseconds by which the endpoint will delay sending acknowledgments.

Comment (maskit) [reply iocore/net/quic/QUICAckFrameCreator.cc]: Either `max_ack_delay_out` or `max_ack_delay_in` can be correct and incorrect. It depends on who initiated the connection. QUICNetVC or something else need to pass an appropriate setting y checking connection direction.

File: iocore/net/QUICNetVConnection.cc [hunk 4]
```diff
@@ -284,8 +284,10 @@ QUICNetVConnection::start()
   this->_frame_dispatcher->add_handler(this->_path_validator);
   this->_frame_dispatcher->add_handler(this->_handshake_handler);
 
+  this->_ack_frame_manager.set_max_ack_delay(params->max_ack_delay_out());
+
   ink_release_assert(this->thread != nullptr);
-  this->_ack_manager_periodic = this->thread->schedule_every(this, params->max_ack_delay_in(), QUIC_EVENT_ACK_PERIODIC);
+  this->_ack_manager_periodic = this->thread->schedule_every(this, params->max_ack_delay_out(), QUIC_EVENT_ACK_PERIODIC);
```
Comment (maskit) [inline iocore/net/QUICNetVConnection.cc]: However the original code has the same bug, we need to use different settings for incoming connection and outgoing connection.

**Pull request merged**